### DISCUSSION TITLE
How to change the name of generated classNames and css variables

### DISCUSSION THREAD

**Discussion by roblotter** [vanilla-extract-css/vanilla-extract#234] [Q&A] [answered]
Is there a way to change the name of the generated classnames and css variables?
Like how css-modules has `localIdentName` or `getLocalIdent` function.

I want to apply my own hash/name/prefix to classnames and css variables and haven't found it in the docs.

Thanks

**Comment by mattcompiles** [comment]
Currently there is no way to do this. What problem do you need to solve that isn't ideal with the current solution?

**Reply by roblotter** [reply]
Well, I think that's the whole point of CSS variables being generated by vanilla-extract, so they can be easily overridden.

Very useful for building reusable components and component libraries.
Same goes for classNames, building a reusable component needs stable classNames - for those scenarios where you want to expose not only CSS variables but some classnames as well

**Reply by roblotter** [reply]
Also related to https://github.com/seek-oss/vanilla-extract/discussions/222

**Reply by mryechkin** [reply]
Wondering if this has been given more consideration, or if anyone has found an alternative solution? Currently investigating `vanilla-extract` to replace our existing styling solution in a design system component library we're building. Most consuming applications will be in React, however there are some cases where apps are simple HTML/CSS, and could benefit from consuming the same styles as the React components from the design system. It would be really great to be able to define styles once, and have some way of generating non-hashed, human-readable CSS classes. This way both React and non-React applications can still utilize the design system specifications.

**Reply by roblotter** [reply]
I think the stable CSS variable names part of this discussion was address by https://github.com/seek-oss/vanilla-extract/discussions/313#discussioncomment-1225189
 and by implementing createGlobalThemeContract.
Non-hashed classes I think is not a feature to be added to vanilla extract - that's basically the whole point of it as a successor to CSS Modules.

**Reply by plesiecki** [reply]
I would use custom namer function to migrate existing CSS Modules design system into vanilla-extract.

**Reply by altrim** [reply]
@mattcompiles Any updates after the discussions? Are there any plans to add support for this?

**Reply by loliver** [reply]
> ... [#313 (comment)](https://github.com/vanilla-extract-css/vanilla-extract/discussions/313#discussioncomment-1225189) and by implementing createGlobalThemeContract ...

Any chance `createGlobalThemeContract` could get an option to set only the `className` generated by the theme? Alternatively for the hashed `className` to always be the same as long as the contents of the theme are the same?

**Reply by loliver** [reply]
So I wrote this little hack after messing around with the `processCss` option in the esbuild plugin which will serve this purpose for us for the time being.

```js
const processCss = async (css) => {
  if (css.match(/--libraryname-colorScheme: light/)) {
    css = `.libraryname-light-theme {
${css.split('{')[1]}`
  }
  if (css.match(/--libraryname-colorScheme: dark/)) {
    css = `.libraryname-dark-theme {
${css.split('{')[1]}`
  }
  return css
}
```

**Comment by graup** [comment]
My team is also interested in this feature, mostly for future-proofing. Imagine some day we stop using our component system or even JS altogether but we still want to use the CSS styles. It would be pretty cumbersome with the current way classes are named. This is the "use vanilla-extract to make your own tailwind" use case. @markdalgleish

**Comment by adregia** [comment]
This has actually become problematic for a project I'm on that leverages Webpack Module Federation. We have a component library using vanilla-extract, and the parent and child modules may need different versions of that library. Because of the inability to change classnames, classes can often intersect between different versions and result in bugs or unexpected behaviors when the modules don't share the same version of the lib.

**Reply by jgornick** [reply]
^^^ This is a great point.

Having the ability to provide a constant class name would be beneficial here for sure.

**Reply by altrim** [reply]
+1 on this. 
We are having the same problem with Module Federation. Were using CSS Modules with custom `getLocalIdent` that took into account the lib version and that worked great. Now with vanilla-extract we are facing the issue where classes intersect between versions, would be nice to get an API to customise the identifiers.

**Comment by jgornick** [comment]
Is there a way we could use a decorator in the code to name the class? Or a comment with metadata specific to vanilla-extract?

**Comment by morbidick** [comment]
To add another usecase: we are planning on using VE in a Microfrontend environment, so multiple UIs will generate their own css file. To prevent classes from colliding it would be necessary to add a prefix to classes for each microfontend.

**Comment by lstephensca** [comment]
I currently have a usecase for this functionality as well. We basically have 2 different workflows:

1. Plain html utilizing a custom css framework for rapid development
2. More in depth one-off projects where js/ts can be utilized with custom css (usually scss)

I though I might give VE a go to develop the css framework as it would make managing the codebase much easier and a bit easier to update. My current workflow is using vite to ingest the scss (many @use/@forward being used), and I get a nice bundled css file that is easy to read/use for the vanilla html/css projects. However, as it stands, the class names/hashes are just too hard to remember/utilize. I could use VE for the second item in my list, but then I would need to also develop a standalone solution for the first which isn't feasible.

**Comment by CroModder** [comment, marked as answer]
Since I found this Q&A and discussion that's not answered, if anyone stubbles on this - it is now possible to change hash identifiers: https://github.com/vanilla-extract-css/vanilla-extract/pull/1160
